Add native support for using a bulk renaming tool.

mtwebster opened the pull request:
This allows the normal method of renaming a file launch a bulk
renamer if one is configured and multiple files are selected.

Ported from nemo.

mbkma commented:
Thanks! It is very handy to have bulk rename built in. However, a few comments:
 - I do not think it is a good idea that the user has to enter a command manually in caja-preferences. The bulk rename tool should be automatically detected by the code, or hardcoded (such as in mate-panel window list applet).
 - It is currently possible to enter any command, so e.g. `atril` will open two instances of atril instead of renaming. This sould be avoided.
 
I tested it with the thunar bulk renamer ( `thunar --bulk-rename` ) it works fine.

mtwebster commented:
Hi, thanks for the feedback -

1 - I'm not sure why we shouldn't allow the user to define the bulk renamer - this allows them to use whichever one they prefer - there are a number of them out there. A default can always be defined in the settings gschema file, as well as overridden by distributions. At the very least the setting should exist, even if you'd prefer not to have it in the preferences ui.

2 - We can't detect what is a bulk renamer and what isn't. Of course the user can put anything in there, but I think you have to give them some benefit of the doubt and assume they won't enter something in there that *isn't* actually a bulk renamer.

Thank you

mbkma commented:
Yep, I think just a gsettings key with a default is perfect. No need to expose this to the user in the preferences ui.

For convenience you could try to find some hard coded bulk-rename-tools. Maybe sth like this works:
```c
static const char* bulk_rename_tools[] = {
	"bulky",
        ...
};

int i;
gchar *progname;
for (i = 0; i < G_N_ELEMENTS(bulk_rename_tools); i++)
{
    if ((progname = g_find_program_in_path(bulk_rename_tools[i])) != NULL);
    {
         g_free(progname);
         break;
    }
    g_free(progname);
}

g_settings_set(settings, "bulk-rename-tool", "^ay", bulk_rename_tools[i]);

```

mtwebster commented:
I removed the preferences ui portion of this, and added some hardcoded renamers to check in addition to the gsettings key. If there are any popular ones I missed let me know and I'll add them.

mbkma commented:
Thanks a lot! In a quick test everything works fine.

mbkma commented on src/file-manager/fm-directory-view.c:
```diff
@@ -6421,6 +6421,87 @@ real_action_redo (FMDirectoryView *view)
+        g_free (quoted_parameter);
+    }
+
+        guint i = 0;
```
this `i` can be in the for-loop

mtwebster replied:
I cleaned up the legacy code a bit (I should in nemo also), though I'm not really keen on declaring i in the loop init - it's not consistent with the rest of caja (and personally I think it worsens readability, but I'm old and blind).

lukefromdc commented:
What is the status of this given the issues Mint had with Bulky?

mtwebster commented:
That issue isn't relevant - it was a problem with the pinned caja version in Mint, and its hardcoded use of bulky. I reviewed this PR since then to be sure, but it already guards against that problem.

lukefromdc commented:
What is the status of this? Are people still interested in working on it, should we close it, or something else?

lukefromdc commented:
Closing as no response from originator in over a year. We can reopen this if folks want to finish it

Pikku-a commented:
> What is the status of this? Are people still interested in working on it, should we close it, or something else?

> Closing for now as no response from originator in over a year. We can reopen this if folks want to finish it

May I ask what changes would be necessary for this pull request?

lukefromdc commented:
I really don't know, all I did on this was some testing as I've never used such a tool

lukefromdc commented:
Reopening as someone might be interested in working on it

lukefromdc commented:
Comments by others would need to be resolved

Pikku-a commented:
I tested this a bit and it seems to work well. At least I didn't notice anything to improve.

mtwebster commented:
The PR probably works because it's pointed at an older version of caja.

Quite a bit of refactoring has occurred since and it became no longer trivial to rebase this PR. I don't have any spare time to work on it currently.

lukefromdc commented:
OK, I'll close this for now. It can be reopened if folks want to work in it in the future, or a new PR can be submitted if major code changes from this are needed
